### PR TITLE
Fix #9103: Fix implicit scopes

### LINKED ISSUE #9094: Non-definition prefix is in implicit scope, unlike Scala 2
## Minimized code

```scala
trait X {
  trait S
}

class Xa(val internal: X) {
  type S = internal.S
  def s: S = new S{}

  implicit class XS(s: S) {
    def printX: Unit = println(s)
  }
}

object Xa extends Xa(new X{})

object App extends App {
  Xa.s.printX
}
```

## Output

```scala
    Xa$$anon$1@55924a6a
```

The above, unexpectedly, compiles and works with Dotty[[1]](https://scastie.scala-lang.org/2DosmGcVTBa8T3TnSqGInQ), but not with Scala 2.13[[2]](https://scastie.scala-lang.org/jtnGsXFHTIGIxoHWdyl2zQ)

## Expectation

It seems like `Xa` is included in the implicit scope of type `Xa.internal.S` where `Xa` is not a _direct prefix_ of `S`, but is instead only a prefix of its prefix, whereas Scala 2 does not include prefixes of prefixes in implicit scope, but only non-prefix parts of the direct prefix.

### LINKED ISSUE #9103: Ambiguous implicit arguments when referencing an object via a val
## Minimized code

```Scala
object a:
  type Foo[T]
  given as Foo[Unit] = ???

val b = a

def Test = summon[b.Foo[Unit]]
```

## Output

```scala
-- Error: /Users/kmetiuk/Projects/scala3/playground/ambiguous/Test.scala:7:30 --
7 |def Test = summon[b.Foo[Unit]]
  |                              ^
  |ambiguous implicit arguments: both method given_Foo_Unit in object a and method given_Foo_Unit in object a match type b.Foo[Unit] of parameter x of method summon in object DottyPredef
1 error found
```

## Expectation
Should compile. The following examples all compile successfully:

```scala
object a:
  type Foo[T]
  given as Foo[Unit] = ???

def Test = summon[a.Foo[Unit]]
```

```scala
object a:
  type Foo
  given as Foo = ???

val b = a

def Test = summon[b.Foo]
```

### PR DESCRIPTION
**Pull request by odersky**
Clarify implicit scope rules and implement accordingly.

Fixes #9103
Fixes #9094

### CODE REVIEW

**Comment by odersky** [comment]
@milessabin I had to disable two places in shapeless that failed after the changes in this PR. Once was the derives clause for `Pure` in `Opt` (other derives clauses of `Pure` work). The other was a test the tries to summon `Pure[Clist]`. I tried to track it down further but failed. Essentially, the previous implementation of type scope contained some underspecified behavior in the way it made use of the `namedPartsWith` call. That's now cleaned up, with a clear spec what implicit scope is (see updated in `implicit-resolution.md` and an implementation that follows the spec closely. Unfortunately, these cleanups caused the two failures. Can you take a look to see what goes wrong?

 Thanks!

 - Martin

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
@milessabin I might have found the issue: The missing elements are associated with match types. I'll see whether I can fix that.

**Comment by odersky** [comment]
All good now. Shapeless compiles without changes.

**Comment by odersky** [comment]
@anatoliykmetyuk It would be good to get this into 0.25, since it changes search implicit behavior, could potentially introduce errors in code using the old behavior.

File: compiler/src/dotty/tools/dotc/typer/Implicits.scala [hunk 1]
```diff
@@ -473,173 +474,226 @@ object Implicits {
...
+
+    def recur(tp: Type): OfTypeImplicits =
+      val (parts, provisional) = collectParts(tp)
+      val companions = collectCompanions(tp, parts)
+      val result = OfTypeImplicits(tp, companions)(runContext)
+      if Config.cacheImplicitScopes
+         && tp.hash != NotCached
+        && !provisional
+        && (tp eq rootTp)              // first type traversed is always cached
+           || !incomplete.contains(tp) // other types are cached if they are not incomplete
```
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/typer/Implicits.scala:612]: A pair of parentheses are missing here, which means this last condition is top-level, that is: all types that are not incomplete are cached, even if:
* Config.cacheImplicitScopes is false, or
* the types hash is NotCached, or
* the type is provisional

**Pull request merged**